synapse/config/room_directory.py

```diff
@@ -23,70 +23,132 @@ def read_config(self, config):
         alias_creation_rules = config["alias_creation_rules"]
 
         self._alias_creation_rules = [
-            _AliasRule(rule)
+            _RoomDirectoryRule("alias_creation_rules", rule)
             for rule in alias_creation_rules
         ]
 
+        room_list_publication_rules = config["room_list_publication_rules"]
+
+        self._room_list_publication_rules = [
+            _RoomDirectoryRule("room_list_publication_rules", rule)
+            for rule in room_list_publication_rules
+        ]
+
     def default_config(self, config_dir_path, server_name, **kwargs):
         return """
         # The `alias_creation` option controls who's allowed to create aliases
         # on this server.
         #
         # The format of this option is a list of rules that contain globs that
-        # match against user_id and the new alias (fully qualified with server
-        # name). The action in the first rule that matches is taken, which can
-        # currently either be "allow" or "deny".
+        # match against user_id, room_id and the new alias (fully qualified with
+        # server name). The action in the first rule that matches is taken,
+        # which can currently either be "allow" or "deny".
+        #
+        # Missing user_id/room_id/alias fields default to "*".
         #
         # If no rules match the request is denied.
         alias_creation_rules:
-            - user_id: "*"
-              alias: "*"
+            - user_id: "*"    # Matches agaisnt the creator of the alias
+              alias: "*"      # Matches against the alias being created
+              room_id: "*"    # Matches against the room ID the alias is being
+                              # pointed at
+              action: allow
+
+        # The `room_list_publication_rules` option controls who can publish and
+        # which rooms can be published in the public room list.
+        #
+        # The format of this option is the same as that for
+        # `alias_creation_rules`.
+        #
+        # If the room has one or more aliases associated with it, only one of
+        # the aliases needs to match the alias rule. If there are no aliases
+        # then only rules with `alias: *` match.
+        room_list_publication_rules:
```

Comment:
I think we're adopting a convention that new sections should be commented out in the default config, which apart from anything else answers the question of "does it matter if I don't include this section in my config". See https://github.com/matrix-org/synapse/blob/develop/synapse/config/server.py#L461 etc.

Reply:
Hmm, ok. It's somewhat annoying to have to comment about the config in the default config and have it in the `read_config`. FWIW what's in the default_config *is* the default if you omit the key